File: Ships.py
class Ships:
    def isShipped(self,arr,mid,D):
        totalDays = 1
        currentWeight = 0
        for i in arr:
            if currentWeight+i > mid:
                totalDays+=1
                currentWeight = i
            else:
                currentWeight+=i
        return True if totalDays<=D else False
    def leastWeightCapacity(self, arr, D):
        low = 0
        high = 0
        for i in arr:
            low = max(low,i)
            high += i
        while low<=high:
            mid = (low+high)//2
            if self.isShipped(arr,mid,D):
                high = mid-1
            else:
                low = mid+1
        return low
arr = list(map(int,input().split()))
D = int(input())
obj = Ships()
result = obj.leastWeightCapacity(arr, D)
print(result)

File: Ships.java
import java.util.Scanner;

public class Ships {
    public static boolean isShipped(int []arr, int days, int weight){
        int totalDays = 1,currentWeight = 0;
        for(int element:arr){
            if(currentWeight+element > weight){
                totalDays+=1;
                currentWeight=element;
            }
            else{
                currentWeight+=element;
            }
        }
        return totalDays <= days;
    }
    static int leastWeightCapacity(int[] arr, int n, int d) {
        int low = 0, high = 0;
        for(int element : arr){
            high += element;
            low = Math.max(low,element);
        }
        while(low <= high){
            int mid = low+(high-low)/2;
            if(isShipped(arr,d,mid)) high = mid - 1;
            else low = mid+1;
        }
        return low;
    }
    public static void main(String[] args){
        Scanner scanner = new Scanner(System.in);
        String[] input = scanner.nextLine().split(" ");
        int[] arr = new int[input.length];
        for(int i=0;i<arr.length;i++){
            arr[i] = Integer.parseInt(input[i]);
        }
        int k = scanner.nextInt();
        int m = scanner.nextInt();
        Ships obj = new Ships();
        System.out.print("Result:"+obj.leastWeightCapacity(arr,k,m));
    }
}
